tests/unit/renderers/0-3-test.js

```diff
@@ -0,0 +1,622 @@
+    ]
+  };
+  let unknownCardHandler = ({env, payload, options}) => {
+    assert.equal(env.name, cardName, 'correct name');
+    assert.ok(!env.isInEditor, 'correct isInEditor');
+    assert.ok(!!env.onTeardown, 'onTeardown hook exists');
+
+    assert.deepEqual(payload, expectedPayload, 'correct payload');
+    assert.deepEqual(options, cardOptions, 'correct options');
+  };
+  renderer = new Renderer({cards: [], unknownCardHandler, cardOptions});
+  renderer.render(mobiledoc);
+});
+
+test('throws if given an object of cards', (assert) => {
+  let cards = {};
+  assert.throws(
+    () => { new Renderer({cards}) }, // jshint ignore: line
+    /`cards` must be passed as an array/
+  );
+});
+
+test('teardown hook calls registered teardown methods', (assert) => {
+  let didTeardown;
+  let card = {
+    name: 'ok',
+    type: 'text',
+    render({env}) {
+      env.onTeardown(() => didTeardown = true);
+    }
+  };
+  let mobiledoc = {
+    version: MOBILEDOC_VERSION,
+    atoms: [],
+    cards: [
+      [card.name]
+    ],
+    markups: [],
+    sections: [
+      [CARD_SECTION_TYPE, 0]
+    ]
+  };
+  renderer = new Renderer({cards: [card]});
+  let { teardown } = renderer.render(mobiledoc);
+
+  assert.ok(!didTeardown, 'precond - no teardown');
+
+  teardown();
+
+  assert.ok(didTeardown, 'teardown callback called');
+});
+
+test('throws when given an unexpected mobiledoc version', (assert) => {
+  let mobiledoc = {
+    version: '0.1.0',
+    atoms: [],
+    cards: [],
+    markups: [],
+    sections: []
+  };
+  assert.throws(
+    () => renderer.render(mobiledoc),
+    /Unexpected Mobiledoc version.*0.1.0/);
+
+  mobiledoc.version = '0.2.1';
+  assert.throws(
+    () => renderer.render(mobiledoc),
+    /Unexpected Mobiledoc version.*0.2.1/);
+});
+
+test('renders a mobiledoc with multiple sections', (assert) => {
+  let mobiledoc = {
+    version: MOBILEDOC_VERSION,
+    atoms: [],
+    cards: [],
+    markups: [],
+    sections: [
+      [MARKUP_SECTION_TYPE, 'P', [
+        [MARKUP_MARKER_TYPE, [], 0, 'first section'],
+      ]],
+      [MARKUP_SECTION_TYPE, 'P', [
+        [MARKUP_MARKER_TYPE, [], 0, 'second section']
+      ]]
+    ]
+  };
+
+  let {result: rendered} = renderer.render(mobiledoc);
+  assert.equal(rendered, ['first section', 'second section'].join('\n'));
+});
+
+test('XSS: unexpected markup and list section tag names are not renderered', (assert) => {
+  let mobiledoc = {
+    version: MOBILEDOC_VERSION,
+    atoms: [],
+    cards: [],
+    markups: [],
+    sections: [
+      [MARKUP_SECTION_TYPE, 'script', [
+        [MARKUP_MARKER_TYPE, [], 0, 'alert("markup section XSS")']
+      ]],
+      [LIST_SECTION_TYPE, 'script', [
+        [[MARKUP_MARKER_TYPE, [], 0, 'alert("list section XSS")']]
+      ]]
+    ]
+  };
+  let { result } = renderer.render(mobiledoc);
+  assert.ok(result.indexOf('script') === -1, 'no script tag rendered');
+});
+
+test('XSS: unexpected markup types are not rendered', (assert) => {
```

Comment:
Yah, which were cut/paste from the HTML renderer :-) :turtle: :turtle: :turtle: :turtle: :turtle: all the way down.